core/trino-main/src/main/java/io/trino/operator/join/unspilled/JoinProbe.java

```diff
@@ -108,23 +106,85 @@ public Page getPage()
+            if (nonNullCount < positionCount) {
+                // We only store positions that are not null
+                int[] positions = new int[nonNullCount];
+                nonNullCount = 0;
+                for (int i = 0; i < positionCount; i++) {
+                    if (!isNull[i]) {
+                        positions[nonNullCount] = i;
+                    }
+                    // This way less code is in the if branch and CPU should be able to optimize branch prediction better
+                    nonNullCount += isNull[i] ? 0 : 1;
+                }
+                long[] packedPositionCache;
+                if (probeHashBlock != null) {
+                    long[] hashes = new long[nonNullCount];
+                    for (int i = 0; i < nonNullCount; i++) {
+                        hashes[i] = BIGINT.getLong(probeHashBlock, positions[i]);
+                    }
+                    packedPositionCache = lookupSource.getJoinPosition(positions, probePage, page, hashes);
+                }
+                else {
+                    packedPositionCache = lookupSource.getJoinPosition(positions, probePage, page);
+                }
+                // Unpack
+                nonNullCount = 0;
+                for (int i = 0; i < positionCount; i++) {
+                    if (!isNull[i]) {
+                        joinPositionCache[i] = packedPositionCache[nonNullCount];
+                    }
+                    // This way less code is in the if branch and CPU should be able to optimize branch prediction better
+                    nonNullCount += isNull[i] ? 0 : 1;
+                }
+                return joinPositionCache;
+            } // else fall back to non-null path
+        }
+        int[] positions = new int[positionCount];
+        for (int i = 0; i < positionCount; i++) {
+            positions[i] = i;
+        }
+        if (probeHashBlock != null) {
+            long[] hashes = new long[positionCount];
+            for (int i = 0; i < positionCount; i++) {
+                hashes[i] = BIGINT.getLong(probeHashBlock, i);
             }
+            return lookupSource.getJoinPosition(positions, probePage, page, hashes);
+        }
+        else {
+            return lookupSource.getJoinPosition(positions, probePage, page);
         }
-        return false;
     }
 
-    private static boolean probeMayHaveNull(Page probePage)
+    private static int getIsNull(List<Block> nullableBlocks, int positionCount, Page probePage, boolean[] isNull)
     {
-        for (int i = 0; i < probePage.getChannelCount(); i++) {
-            if (probePage.getBlock(i).mayHaveNull()) {
-                return true;
+        for (int i = 0; i < nullableBlocks.size() - 1; i++) {
+            Block block = probePage.getBlock(i);
+            for (int j = 0; j < positionCount; j++) {
+                isNull[j] |= block.isNull(j);
             }
         }
-        return false;
+        // Last block will also calculate `nonNullCount`
```

Comment:
one thing worth checking (if you haven't already) is leaving the last block with the others and doing a separate loop for `nonNullCount`.
then both loops are highly _jit optimizable_ (first is just ORing the byte arrays, second is a sum of bytes)

Reply:
This would usually end up being only the last loop since most joins are single column. I tried different approaches and this one worked best.